### ISSUE TITLE
allow for startup scripts

### ISSUE THREAD

**Issue by javamonkey79** [neo4j/docker-neo4j#166]
**Scenario Steps**
1. load neo4j, using the admin tool
2. start neo4j using docker

**Missing Feature**
1. I need indexes on my data
2. Given the last thing the neo4j dockerfile does, is startup neo4j, I must add indexes manually. 

**Suggested Fix**
1. Create a new EXTENSION_SCRIPT variable in docker-entrypoint.sh that is ran at the end, perhaps named STARTUP_EXTENSION_SCRIPT

**Open Questions**
1. Should EXTENSION_SCRIPT be renamed? What is the best name for STARTUP_EXTENSION_SCRIPT?
2. Would a startup script need to be ran after console mode?
3. Would any additional code added to the end of docker-entrypoint.sh be executed, or, does the script block? 

**Related Posts**
1. [This stackoverflow post](https://stackoverflow.com/questions/48357238/how-can-i-run-cypher-scripts-on-startup-of-a-neo4j-docker-container)

**Comment by javamonkey79** [comment]
OK, I was able to use the documentation [here](https://docs.docker.com/config/containers/multi-service_container/) to create a wrapper script, and an indexes script that just sleeps until neo4j is back up, like so:

Dockerfile
```
FROM neo4j:latest

ENV NEO4J_AUTH=neo4j/password

RUN apk add --no-cache --quiet procps

COPY create-indexes.sh create-indexes.sh
COPY wrapper.sh wrapper.sh

ENTRYPOINT ["./wrapper.sh"]
```

wrapper.sh:
```
#!/bin/bash

# turn on bash's job control
set -m

# Start the primary process and put it in the background
/docker-entrypoint.sh neo4j &

# Start the helper process
./create-indexes.sh

# the my_helper_process might need to know how to wait on the
# primary process to start before it does its work and returns


# now we bring the primary process back into the foreground
# and leave it there
fg %1
```

create-indexes.sh
```
#!/bin/bash

until cypher-shell -u neo4j -p shaun123 'CREATE INDEX ON :Page(url);'
do
  echo "create page index failed, sleeping"
  sleep 10
done

until cypher-shell -u neo4j -p shaun123 'CREATE INDEX ON :Visited(url);'
do
  echo "create visited index failed, sleeping"
  sleep 10
done

until cypher-shell -u neo4j -p shaun123 'CREATE INDEX ON :Anchor(url);'
do
  echo "create anchor index failed, sleeping"
  sleep 10
done
```

So, I guess this feature is no longer needed (and probably not even possible).

**Comment by roschaefer** [comment]
Could we re-open this issue, please? I understand the solution @javamonkey79 but you hard-code the username and password into `create-indexes.sh`. If I build a customized docker image and want to publish it to dockerhub, this is unacceptable.

Is there any way how to setup indices and constraints in a `Dockerfile` **without** creating a user?

**Comment by javamonkey79** [comment]
Mine was just an example. You don't have to provide user unless you have a user set and need to do some operation with it (such as create an index). 

Also, un + pw is not hard coded in my actual working version, but, parameterized with env vars.

**Comment by javamonkey79** [comment]
Here is a newer version of my index script:
```
#!/bin/bash

create_index()
{
	echo "CREATE INDEX ON :$1($2);"
	until cypher-shell -u "neo4j" -p $NEO4J_SECRETS_PASSWORD "CREATE INDEX ON :$1($2);"
	do
	  echo "create index on ($1) failed, sleeping"
	  sleep 10
	done
	echo "create index on ($1) complete"
}

create_index "Page" "url"
create_index "Visited" "url"
create_index "NonVisited" "url"
```

`NEO4J_SECRETS_PASSWORD` can just be passed to the docker image on run. Also, you could introduce a check in this script, to see if env vars are set, and if so use them, if not assume that it is unsecured, or, let it error out.

**Comment by sfuerte** [comment]
@javamonkey79, you may want to add some wait period in `wrapper.sh` before calling to `./create-indexes.sh`: 
```
# wait for Neo4j
while true; do
    sleep 5
    if curl -s -I http://localhost:7474 | grep -q "200 OK"; then
        echo "Ready"
        break
    else
        echo "Not Ready"
        continue
    fi
done
```

**Comment by ticapix** [comment]
> 
> 
> @javamonkey79, you may want to add some wait period in `wrapper.sh` before calling to `./create-indexes.sh`:
> 
> ```
> # wait for Neo4j
> while true; do
>     sleep 5
>     if curl -s -I http://localhost:7474 | grep -q "200 OK"; then
>         echo "Ready"
>         break
>     else
>         echo "Not Ready"
>         continue
>     fi
> done
> ```

more compact solution with `wget` which is part of the `alpine` image used by `neo4j`

```shell
wget --tries=10 --waitretry=2 -O /dev/null http://localhost:7474
```

my final `wrapper.sh` script used for testing

```shell
#!/bin/sh

export NEO4J_USERNAME=$NEO4J_USERNAME
export NEO4J_PASSWORD=$NEO4J_PASSWORD

set -e # exit at the first error

# Start the primary process and put it in the background
/docker-entrypoint.sh neo4j &

# wait for Neo4j
wget --quiet --tries=10 --waitretry=2 -O /dev/null http://localhost:7474

for f in /cyphers/*.cyp; do
    echo "running cypher ${f}"
    cypher-shell --file $f
done

cypher-shell --format plain "MATCH (n) RETURN count(n) AS count"
```

**Comment by GregCKrause** [comment]
> > @javamonkey79, you may want to add some wait period in `wrapper.sh` before calling to `./create-indexes.sh`:
> > ```
> > # wait for Neo4j
> > while true; do
> >     sleep 5
> >     if curl -s -I http://localhost:7474 | grep -q "200 OK"; then
> >         echo "Ready"
> >         break
> >     else
> >         echo "Not Ready"
> >         continue
> >     fi
> > done
> > ```
> 
> more compact solution with `wget` which is part of the `alpine` image used by `neo4j`
> 
> ```shell
> wget --tries=10 --waitretry=2 -O /dev/null http://localhost:7474
> ```
> 
> my final `wrapper.sh` script used for testing
> 
> ```shell
> #!/bin/sh
> 
> export NEO4J_USERNAME=$NEO4J_USERNAME
> export NEO4J_PASSWORD=$NEO4J_PASSWORD
> 
> set -e # exit at the first error
> 
> # Start the primary process and put it in the background
> /docker-entrypoint.sh neo4j &
> 
> # wait for Neo4j
> wget --quiet --tries=10 --waitretry=2 -O /dev/null http://localhost:7474
> 
> for f in /cyphers/*.cyp; do
>     echo "running cypher ${f}"
>     cypher-shell --file $f
> done
> 
> cypher-shell --format plain "MATCH (n) RETURN count(n) AS count"
> ```

This leaves the main process in the background. I think @javamonkey79's example of `fg` is the best approach here, as it moves the intended main process back to the foreground.

**Comment by fipro78** [comment]
The `wrapper.sh` script of @ticapix has two issues:
- the `wget` call will not perform retries, because the error is "connection refused" which avoids the retry as written in the man pages
- if there are no files in the directory, the pattern will resolve to itself, which leads to a "No such file or directory" error

The fixed version with the `fg` approach looks like this

```
#!/bin/bash

export NEO4J_USERNAME=$NEO4J_USERNAME
export NEO4J_PASSWORD=$NEO4J_PASSWORD

# turn on bash's job control
set -m

# Start the primary process and put it in the background
/startup/docker-entrypoint.sh neo4j &

# wait for Neo4j
wget --quiet --tries=10 --retry-connrefused=on --waitretry=5 -O /dev/null http://localhost:7474

# Allows patterns which match no files to expand to a null string, rather than themselves
# Avoids (No such file or directory) if no .cql files are in the logs directory
shopt -s nullglob

# Start the helper process
for f in /cyphers/*.cyp; do
    echo "running cypher ${f}"
    cypher-shell --file $f
done

# unset nullglob
shopt -u nullglob

# now we bring the primary process back into the foreground
# and leave it there
fg %1
```

Just in case someone stumbles over this ticket in the future and wonders why it does not directly work.
